File: samples/export_batch-t/C#/StructorizerFileAPI.cs
﻿/*
	Structorizer
	A little tool which you can use to create Nassi-Shneiderman Diagrams (NSD)

	Copyright (C) 2009, 2020  Bob Fisch

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or any
	later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
 
/******************************************************************************************************
 *
 *	  Author:		Kay Gürtzig
 *
 *	  Description:	File API for the CSharpGenerator.
 *
 ******************************************************************************************************
 *
 *	  Revision List
 *
 *	  Author          Date        Description
 *	  ------          ----        -----------
 *	  Kay Gürtzig     2017-01-04  First Issue
 *    Kay Gürtzig     2017-01-05  Functional improvements (in StructorizerFileAPI class)
 *
 ******************************************************************************************************
 *
 *	  Comment:
 *	  
 *
 ******************************************************************************************************/// 

namespace FileAPI.CS
{
	class StructorizerFileAPI
	{
		// Map of open stream readers
		private static Dictionary<int, StreamReader> readers = new Dictionary<int, StreamReader>();
		// Map of open stream writers
		private static Dictionary<int, StreamWriter> writers = new Dictionary<int, StreamWriter>();
		// Map of the current lines of the respective open stream readers
		private static Dictionary<int, string> lines = new Dictionary<int, string>();
		// Next usable file number (handle)
		private static int nextFileNo = 1;

		/**
		 * Tries to open a text file with given filePath for reading. File must exist.
		 * A negative or zero return value indicates failure.
		 * @param filePath - the path of the file (may be absolute or relative to the current directory)
		 * @return an int to be used as file handle for this API if > 0 or as error code otherwise.
		 */
		public static int fileOpen(string filePath)
		{
			int fileNo = 0;
			try {
				StreamReader reader = new StreamReader(filePath);
				fileNo = nextFileNo;
				readers.Add(nextFileNo++, reader);
			}
			catch (FileNotFoundException) {
				fileNo = -2;
			}
			catch (IOException) {
				fileNo = -1;
			} 
			return fileNo;
		}

		/**
		 * Tries to create a text file with given filePath for writing. Is file exists then it will
		 * be cleared (without warning!).
		 * A negative or zero return value indicates failure.
		 * @param filePath - the path of the file (may be absolute or relative to the current directory)
		 * @return an int to be used as file handle for this API if > 0 or as error code otherwise.
		 */
		public static int fileCreate(string filePath)
		{
			int fileNo = 0;
			try {
				StreamWriter writer = new StreamWriter(filePath);
				fileNo = nextFileNo;
				writers.Add(nextFileNo++, writer);
			}
			catch (UnauthorizedAccessException) {
				fileNo = -3;
			}
			catch (DirectoryNotFoundException) {
				fileNo = -2;
			}
			catch (IOException) {
				fileNo = -1;
			} 
			return fileNo;
		}

		/**
		 * Tries to create or open a text file with given filePath for writing. If the file had existed
		 * then it will not be cleared but writing starts at previous end.
		 * A negative or zero return value indicates failure.
		 * @param filePath - the path of the file (may be absolute or relative to the current directory)
		 * @return an int to be used as file handle for this API if > 0 or as error code otherwise.
		 */
		public static int fileAppend(string filePath)
		{
			int fileNo = 0;
			try {
				StreamWriter writer = new StreamWriter(filePath, true);
				fileNo = nextFileNo;
				writers.Add(nextFileNo++, writer);
			}
			catch (UnauthorizedAccessException) {
				fileNo = -3;
			}
			catch (DirectoryNotFoundException) {
				fileNo = -2;
			}
			catch (IOException) {
				fileNo = -1;
			} 
			return fileNo;
		}

		/**
		 * Closes the file with given fileNo handle. If fileNo is not associated with an open file
		 * then an IOException will be thrown.
		 * @param fileNo - file handle as obtained by fileOpen, fileCreate or fileAppend before
		 */
		public static void fileClose(int fileNo)
		{
			if (readers.ContainsKey(fileNo)) {
				readers[fileNo].Close();
				readers.Remove(fileNo);
			}
			else if (writers.ContainsKey(fileNo)) {
				writers[fileNo].Close();
				writers.Remove(fileNo);
			}
		}

		/**
		 * Checks whether the input file with given fileNo handle is exhausted i.e. provides no
		 * readable content beyond the current reading position.
		 * If fileNo is not associated to any opened readable file then an IOException will be thrown.
		 * @param fileNo - file handle as obtained by fileOpen before.
		 * @return true iff end of file has been reached.
		 */
		public static bool fileEOF(int fileNo)
		{
			bool isEOF = true;
			if (readers.ContainsKey(fileNo)) {
				isEOF = fetchLine(fileNo, true) == null;
			}
			return isEOF;
		}

		/**
		 * Internal helper method.
		 * Provides the current text line (or the rest of the current text line) from the
		 * text file given by fileNo handle without "consuming" it.
		 * May throw an error if the given handle is not associated to an open text input file.
		 * If the file input stream was exhausted (i.e. is at end of file) then null will
		 * be returned.
		 * @param fileNo - file handle as obtained by fileOpen before
		 * @return the current line from file input or null.
		 */
		private static string fetchLine(int fileNo, bool yieldEmptyLine)
		{
			string line = null;
			if (readers.ContainsKey(fileNo))
			{
				if (lines.ContainsKey(fileNo))
				{
					line = lines[fileNo];
				}
				if (line == null || line.Length == 0 && !yieldEmptyLine)
				{
					if (readers[fileNo].EndOfStream)
					{
						line = null;
						lines.Remove(fileNo);
					}
					else
					{
						try
						{
							line = readers[fileNo].ReadLine();
							lines[fileNo] = line;
						}
						catch (IOException)
						{
							line = null;
						}
					}
				}
			}
			return line;
		}

		/**
		 * Reads the next token from the text file given by fileNo handle and returns it as
		 * boxed value of one of the types int, double, or string.
		 * Throws an error if the given handle is not associated to an open text input file.
		 * The result may also be interpreted as bool value in which case false means that the
		 * value could not be interpreted as any of the known types.
		 * @param fileNo - file handle as obtained by fileOpen before
		 * @return wrapper object for the current value as interpreted from file input.
		 */
		public static object fileRead(int fileNo)
		{
			object value = null;
			string line = null;
			do
			{
				line = fetchLine(fileNo, false);
			} while (line != null && line.Trim().Length == 0);
			if (line != null)
			{
				line = line.TrimStart();
				string[] tokens = line.Split(null);
				int intVal;
				double dblVal;
				if (Int32.TryParse(tokens[0], out intVal)) {
					value = intVal;
					line = line.Substring(tokens[0].Length);
				}
				else if (Double.TryParse(tokens[0], out dblVal)) {
					value = dblVal;
					line = line.Substring(tokens[0].Length);
				}
				else {
					string strVal = tokens[0];
					line = line.Substring(tokens[0].Length);
					if (strVal.StartsWith("\"") || strVal.StartsWith("'")) {
						string delimiter = strVal.Substring(0,1);
						for (int i = 1; i < tokens.Length && !(strVal.Length > 1 && strVal.EndsWith(delimiter)); i++) {
							int start = line.IndexOf(tokens[i]);
							strVal += line.Substring(0, start) + tokens[i];
							line = line.Substring(start + tokens[i].Length);
						}
						strVal = strVal.Substring(1);
						if (strVal.EndsWith(delimiter)) {
							strVal = strVal.Substring(0, strVal.Length-1);
						}
					}
					value = strVal;
				}
				if (value != null)
				{
					lines[fileNo] = line;
				}
			}
			return value;
		}

		/**
		 * Reads the next char from the text file given by fileNo handle and returns it.
		 * Throws an error if the given handle is not associated to an open text input file.
		 * @param fileNo - file handle as obtained by fileOpen before
		 * @return the current char from file input char sequence or '\0'.
		 */
		public static char fileReadChar(int fileNo)
		{
			char ch = '\0';
			string line = fetchLine(fileNo, true);
			if (line != null)
			{
				if (line.Length > 0)
				{
					ch = line[0];
					lines[fileNo] = line.Substring(1);
				}
				else if ((line = fetchLine(fileNo, false)) != null)
				{
					ch = '\n';
				}
			}
			return ch;
		}

		/**
		 * Reads the next int value from the text file given by fileNo handle and returns it.
		 * Throws an error if the given handle is not associated to an open text input file.
		 * If the file input stream was exhausted (was at end of file) or if the token at reading
		 * position was not interpretable as integral literal, then null will be returned.
		 * @param fileNo - file handle as obtained by fileOpen before
		 * @return the current int number as interpreted from file input char sequence or null.
		 */
		public static int fileReadInt(int fileNo)
		{
			int value = 0;
			string line = null;
			do
			{
				line = fetchLine(fileNo, false);
			} while (line != null && line.TrimStart().Length == 0);
			if (line != null)
			{
				line = line.TrimStart();
				string[] tokens = line.Split(null);
				value = Int32.Parse(tokens[0]);
				lines[fileNo] = line.Substring(tokens[0].Length);
			}
			return value;
		}

		/**
		 * Reads the next floating-point value from the text file given by fileNo handle and
		 * returns it as double.
		 * Throws an error if the given handle is not associated to an open text input file.
		 * If the file input stream was exhausted (was at end of file) or if the token at reading
		 * position was not interpretable as floating-point literal then null will be returned.
		 * @param fileNo - file handle as obtained by fileOpen before
		 * @return the current floating point-value from file input char sequence or null.
		 */
		public static double fileReadDouble(int fileNo)
		{
			double value = 0;
			string line = null;
			do
			{
				line = fetchLine(fileNo, false);
			} while (line != null && line.TrimStart().Length == 0);
			if (line != null)
			{
				line = line.TrimStart();
				string[] tokens = line.Split(null);
				value = Double.Parse(tokens[0]);
				lines[fileNo] = line.Substring(tokens[0].Length);
			}
			return value;
		}

		/**
		 * Reads the next text line (or the rest of the current text line) from the text file
		 * given by fileNo handle and returns it. This consumes the line.
		 * May throw an error if the given handle is not associated to an open text input file.
		 * If the file input stream had been exhausted (is at end of file) then null will be
		 * returned.
		 * @param fileNo - file handle as obtained by fileOpen before
		 * @return the current line from file input or null.
		 */
		public static string fileReadLine(int fileNo)
		{
			string line = fetchLine(fileNo, true);
			if (line != null)
			{
				// Line is consumed now
				lines.Remove(fileNo);
			}
			return line;
		}

		/**
		 * Writes the given value as char sequence to the file given by handle fileNo.
		 * Throws an error if the given handle is not associated to an open text output file.
		 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
		 * @param data the value to be written to file.
		 */
		public static void fileWrite(int fileNo, int data)
		{
			writers[fileNo].Write(data);
		}
		/**
		 * Writes the given value as char sequence to the file given by handle fileNo.
		 * Throws an error if the given handle is not associated to an open text output file.
		 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
		 * @param data the value to be written to file.
		 */
		public static void fileWrite(int fileNo, double data)
		{
			writers[fileNo].Write(data);
		}
		/**
		 * Writes the given value as char sequence to the file given by handle fileNo.
		 * Throws an error if the given handle is not associated to an open text output file.
		 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
		 * @param data the value to be written to file.
		 */
		public static void fileWrite(int fileNo, string data)
		{
			writers[fileNo].Write(data);
		}
		/**
		 * Writes the given value as char sequence to the file given by handle fileNo.
		 * Throws an error if the given handle is not associated to an open text output file.
		 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
		 * @param data the value to be written to file.
		 */
		public static void fileWrite(int fileNo, char data)
		{
			writers[fileNo].Write(data);
		}
		/**
		 * Writes the given value as char sequence to the file given by handle fileNo.
		 * Throws an error if the given handle is not associated to an open text output file.
		 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
		 * @param data the value to be written to file.
		 */
		public static void fileWrite(int fileNo, bool data)
		{
			writers[fileNo].Write(data);
		}

		/**
		 * Writes the given value as char sequence to the file given by handle fileNo
		 * and appends a newline char or sequence as value separator.
		 * Throws an error if the given handle is not associated to an open text output file.
		 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
		 * @param data the value to be written to file.
		 */
		public static void fileWriteLine(int fileNo, int data)
		{
			writers[fileNo].WriteLine(data);
		}
		/**
		 * Writes the given value as char sequence to the file given by handle fileNo
		 * and appends a newline char or sequence as value separator.
		 * Throws an error if the given handle is not associated to an open text output file.
		 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
		 * @param data the value to be written to file.
		 */
		public static void fileWriteLine(int fileNo, double data)
		{
			writers[fileNo].WriteLine(data);
		}
		/**
		 * Writes the given value as char sequence to the file given by handle fileNo
		 * and appends a newline char or sequence as value separator.
		 * Throws an error if the given handle is not associated to an open text output file.
		 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
		 * @param data the value to be written to file.
		 */
		public static void fileWriteLine(int fileNo, string data)
		{
			writers[fileNo].WriteLine(data);
		}
		/**
		 * Writes the given value as char sequence to the file given by handle fileNo
		 * and appends a newline char or sequence as value separator.
		 * Throws an error if the given handle is not associated to an open text output file.
		 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
		 * @param data the value to be written to file.
		 */
		public static void fileWriteLine(int fileNo, char data)
		{
			writers[fileNo].WriteLine(data);
		}
		/**
		 * Writes the given value as char sequence to the file given by handle fileNo
		 * and appends a newline char or sequence as value separator.
		 * Throws an error if the given handle is not associated to an open text output file.
		 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
		 * @param data the value to be written to file.
		 */
		public static void fileWriteLine(int fileNo, bool data)
		{
			writers[fileNo].WriteLine(data);
		}

	}

}

File: samples/export_batch-t/Java/StructorizerFileAPI.java
/*
    Structorizer
    A little tool which you can use to create Nassi-Schneiderman Diagrams (NSD)

    Copyright (C) 2009  Bob Fisch

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or any
    later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 package lu.fisch.structorizer.generators;

/******************************************************************************************************
 *
 *      Author:         Kay Gürtzig
 *
 *      Description:    File API for the JavaGenerator.
 *
 ******************************************************************************************************
 *
 *      Revision List
 *
 *      Author          Date            Description
 *      ------          ----            -----------
 *      Kay Gürtzig     22.12.2016      First Issue
 *
 ******************************************************************************************************
 *
 *      Comment:
 *      
 *
 ******************************************************************************************************///
 
 public class StructorizerFileAPI 
 {

	/**
	 * Maps the file numbers to stream writers or scanners. For File-API-internal use only 
	 */
	private static java.util.Vector<java.io.Closeable> structorizerFileMap = new java.util.Vector<java.io.Closeable>(); 

	/**
	 * Tries to open a text file with given filePath for reading. File must exist.
	 * A negative or zero return value indicates failure. Failure code explanation:
	 * 0,-1: unspecific error
	 * -2: file not found
	 * -3: insufficient permissions
	 * @param filePath - the path of the file (may be absolute or relative to the current directory)
	 * @return an integer to be used as file handle for this API if > 0 or as error code otherwise.
	 */
	public static int fileOpen(String filePath)
	{
		int fileNo = 0; 
		java.io.File file = new java.io.File(filePath);
		try {
			java.io.FileInputStream fis = new java.io.FileInputStream(file);
			java.io.BufferedReader reader = new java.io.BufferedReader(new java.io.InputStreamReader(fis, "UTF-8"));
			fileNo = structorizerFileMap.size() + 1;
			structorizerFileMap.add(new java.util.Scanner(reader));
		}
		catch (SecurityException e) { fileNo = -3; }
		catch (java.io.FileNotFoundException e) { fileNo = -2; }
		catch (java.io.IOException e) { fileNo = -1; }
		return fileNo;
	}

	/**
	 * Tries to create a text file with given filePath for writing. Is file exists then it will
	 * be cleared (without warning!).
	 * A negative or zero return value indicates failure. Failure code explanation:
	 * 0,-1: unspecific error
	 * -2: path to create the file within not found
	 * -3: insufficient permissions
	 * @param filePath - the path of the file (may be absolute or relative to the current directory)
	 * @return an integer to be used as file handle for this API if > 0 or as error code otherwise.
	 */
	public static int fileCreate(String filePath)
	{
		int fileNo = 0;
		java.io.File file = new java.io.File(filePath);
		try {
			java.io.FileOutputStream fos = new java.io.FileOutputStream(file);
			java.io.BufferedWriter writer = new java.io.BufferedWriter(new java.io.OutputStreamWriter(fos, "UTF-8"));
			fileNo = structorizerFileMap.size() + 1;
			structorizerFileMap.add(writer);
		}
		catch (SecurityException e) { fileNo = -3; }
		catch (java.io.FileNotFoundException e) { fileNo = -2; }
		catch (java.io.IOException e) { fileNo = -1; }
		return fileNo;
	}

	/**
	 * Tries to create or open a text file with given filePath for writing. If the file exists then
	 * it will not be cleared but writing starts at previous end.
	 * A negative or zero return value indicates failure. Failure code explanation:
	 * 0,-1: unspecific error
	 * -2: path to create the file within not found
	 * -3: insufficient permissions
	 * @param filePath - the path of the file (may be absolute or relative to the current directory)
	 * @return an integer to be used as file handle for this API if > 0 or as error code otherwise.
	 */
	public static int fileAppend(String filePath)
	{
		int fileNo = 0;
		java.io.File file = new java.io.File(filePath);
		try {
			java.io.FileOutputStream fos = new java.io.FileOutputStream(file, true);
			java.io.BufferedWriter writer = new java.io.BufferedWriter(new java.io.OutputStreamWriter(fos, "UTF-8"));
			fileNo = structorizerFileMap.size() + 1;
			structorizerFileMap.add(writer);
		}
		catch (SecurityException e) { fileNo = -3; }
		catch (java.io.FileNotFoundException e) { fileNo = -2; }
		catch (java.io.IOException e) { fileNo = -1; }
		return fileNo;
	}

	/**
	 * Closes the file with given fileNo handle. If fileNo is not associated with an open file
	 * then an IOException will be thrown. 
	 * @param fileNo - file handle as obtained by fileOpen, fileCreate or fileAppend before
	 */
	public static void fileClose(int fileNo) throws java.io.IOException
	{
		if (fileNo > 0 && fileNo <= structorizerFileMap.size()) {
			java.io.Closeable file = structorizerFileMap.get(fileNo - 1);
			if (file != null) {
				try { file.close(); }
				catch (java.io.IOException e) {}
				structorizerFileMap.set(fileNo - 1, null); }
		}
		else { throw new java.io.IOException("fileClose: Invalid file number or file not open for reading."); }
	}

	/**
	 * Checks whether the input file with given fileNo handle is exhausted i.e. provides no
	 * readable content beyond the current reading position. 
	 * If fileNo is not associated to any opened readable file then an IOException will be thrown. 
	 * @param fileNo - file handle as obtained by fileOpen before.
	 * @return true iff end of file has been reached.
	 */
	public static boolean fileEOF(int fileNo) throws java.io.IOException
	{
		boolean isEOF = true;
		if (fileNo > 0 && fileNo <= structorizerFileMap.size()) {
			java.io.Closeable reader = structorizerFileMap.get(fileNo - 1);
			if (reader instanceof java.util.Scanner) {
				//try {
					isEOF = !((java.util.Scanner)reader).hasNext();
				//} catch (java.io.IOException e) {}
			}
		}
		else { throw new java.io.IOException("fileEOF: Invalid file number or file not open for reading."); }
		return isEOF;
	}

	/* Internal helper function for file reading */
	private static Object structorizerGetScannedObject(java.util.Scanner sc) {
		Object result = null;
		sc.useLocale(java.util.Locale.UK);
		if (sc.hasNextInt()) { result = sc.nextInt(); }
		else if (sc.hasNextDouble()) { result = sc.nextDouble(); }
		else if (sc.hasNext("\\\".*?\\\"")) {
			String str = sc.next("\\\".*?\\\"");
			result = str.substring(1, str.length() - 1);
		}
		else if (sc.hasNext("'.*?'")) {
			String str = sc.next("'.*?'");
			result = str.substring(1, str.length() - 1);
		}
		else if (sc.hasNext("\\{.*?\\}")) {
			String token = sc.next();
			result = new Object[]{token.substring(1, token.length()-1)};
		}
		else if (sc.hasNext("\\\".*")) {
			String str = sc.next();
			while (sc.hasNext() && !sc.hasNext(".*\\\"")) {
				str += " " + sc.next();
			}
			if (sc.hasNext()) { str += " " + sc.next(); }
			result = str.substring(1, str.length() - 1);
		}
		else if (sc.hasNext("'.*")) {
			String str = sc.next();
			while (sc.hasNext() && !sc.hasNext(".*'")) {
				str += " " + sc.next();
			}
			if (sc.hasNext()) { str += " " + sc.next(); }
			result = str.substring(1, str.length() - 1);
		}
		else if (sc.hasNext("\\{.*")) {
			java.util.regex.Pattern oldDelim = sc.delimiter();
			sc.useDelimiter("\\}");
			String content = sc.next().trim().substring(1);
			sc.useDelimiter(oldDelim);
			if (sc.hasNext("\\}")) { sc.next(); }
			String[] elements = {};
			if (!content.isEmpty()) {
				elements = cotent.split("\\p{javaWhitespace}*,\\p{javaWhitespace}*");
			}
			Object[] objects = new Object[elements.length];
			for (int i = 0; i < elements.length; i++) {
				java.util.Scanner sc0 = new java.util.Scanner(elements[i]);
				objects[i] = structorizerGetScannedObject(sc0);
				sc0.close();
			}
			result = objects;
		}
		else { result = sc.next(); }
		return result;
	}

	/**
	 * Reads the next token from the text file given by fileNo handle and returns it as an
	 * appropriate value of one of the classes Integer, Double, Object[], or String.  
	 * Throws an error if the given handle is not associated to an open text input file. 
	 * @param fileNo - file handle as obtained by fileOpen before
	 * @return the current object as interpreted from file input character sequence or null.
	 */
	public static Object fileRead(int fileNo) throws java.io.IOException
	{
		Object result = null;
		boolean ok = false;
		if (fileNo > 0 && fileNo <= structorizerFileMap.size()) {
			java.io.Closeable reader = structorizerFileMap.get(fileNo - 1);
			if (reader instanceof java.util.Scanner) {
				result = structorizerGetScannedObject((java.util.Scanner)reader);
				ok = true;
			}
		}
		if (!ok) { throw new java.io.IOException("fileRead: Invalid file number or file not open for reading."); }
		return result;
	}

	/**
	 * Reads the next character from the text file given by fileNo handle and returns it.
	 * Throws an error if the given handle is not associated to an open text input file. 
	 * @param fileNo - file handle as obtained by fileOpen before
	 * @return the current character from file input character sequence or '\0'.
	 */
	public static Character fileReadChar(int fileNo) throws java.io.IOException {
		Character result = '\0';
		boolean ok = false;
		if (fileNo > 0 && fileNo <= structorizerFileMap.size()) {
			java.io.Closeable reader = structorizerFileMap.get(fileNo - 1);
			if (reader instanceof java.util.Scanner) {
				java.util.Scanner sc = (java.util.Scanner)reader;
				java.util.regex.Pattern oldDelim = sc.delimiter();
				sc.useDelimiter("");
				try {
					if (!sc.hasNext(".") && sc.hasNextLine()) { sc.nextLine(); result = '\n'; }
					else if (sc.hasNext(".") {
						result = sc.next(".").charAt(0);
					}
				}
				finally { sc.useDelimiter(oldDelim); }
				ok = true;
			}
		}
		if (!ok) { throw new java.io.IOException("fileReadChar: Invalid file number or file not open for reading."); }
		return result;
	}

	/**
	 * Reads the next integer value from the text file given by fileNo handle and returns it.  
	 * Throws an error if the given handle is not associated to an open text input file.
	 * If the file input stream was exhausted (was at end of file) or if the token at reading
	 * was not interpretable as integral literal, then null will be returned. 
	 * @param fileNo - file handle as obtained by fileOpen before
	 * @return the current int number as interpreted from file input character sequence or null.
	 */
	public static Integer fileReadInt(int fileNo) throws java.io.IOException
	{
		Integer result = null;
		boolean ok = false;
		if (fileNo > 0 && fileNo <= structorizerFileMap.size()) {
			java.io.Closeable reader = structorizerFileMap.get(fileNo - 1);
			if (reader instanceof java.util.Scanner) {
				if (((java.util.Scanner)reader).hasNextInt()) {
					result = ((java.util.Scanner)reader).nextInt();
				}
				ok = true;
			}
		}
		if (!ok) { throw new java.io.IOException("fileReadInt: Invalid file number or file not open for reading."); }
		return result;
	}

	/**
	 * Reads the next floating-point value from the text file given by fileNo handle and
	 * returns it as double.  
	 * Throws an error if the given handle is not associated to an open text input file.
	 * If the file input stream was exhausted (was at end of file) or if the token at reading
	 * was not interpretable as integral literal then null will be returned. 
	 * @param fileNo - file handle as obtained by fileOpen before
	 * @return the current floating point-value from file input character sequence or null.
	 */
	public static double fileReadDouble(int fileNo) throws java.io.IOException
	{ 
		Double result = null;
		boolean ok = false;
		if (fileNo > 0 && fileNo <= structorizerFileMap.size()) {
			java.io.Closeable reader = structorizerFileMap.get(fileNo - 1);
			if (reader instanceof java.util.Scanner) {
				if (((java.util.Scanner)reader).hasNextDouble()) {
					result = ((java.util.Scanner)reader).nextDouble();
				}
				ok = true;
			}
		}
		if (!ok) { throw new java.io.IOException("fileReadDouble: Invalid file number or file not open for reading."); }
		return result;
	}

	/**
	 * Reads the next text line (or the rest of the current text line) from the text file
	 * given by fileNo handle and returns it.  
	 * Throws an error if the given handle is not associated to an open text input file.
	 * If the file input stream was exhausted (was at end of file) or if the token at reading
	 * was not interpretable as integral literal then null will be returned. 
	 * @param fileNo - file handle as obtained by fileOpen before
	 * @return the current object as interpreted from file input character sequence or null.
	 */
	public static String fileReadLine(int fileNo) throws java.io.IOException
	{
		String line = null;
		boolean ok = false;
		if (fileNo > 0 && fileNo <= structorizerFileMap.size()) {
			java.io.Closeable reader = structorizerFileMap.get(fileNo - 1);
			if (reader instanceof java.util.Scanner) {
				if (((java.util.Scanner)reader).hasNextLine()) {
					line = ((java.util.Scanner)reader).nextLine();
				}
				ok = true;
			}
		}
		if (!ok) { throw new java.io.IOException("fileReadLine: Invalid file number or file not open for reading."); }
		return line;
	}

	/**
	 * Writes the given value as character sequence to the file given by handle fileNo.  
	 * Throws an error if the given handle is not associated to an open text output file.
	 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
	 * @param data the object or value to be written to file.
	 */
	public static void fileWrite(int fileNo, java.lang.Object data) throws java.io.IOException
	{
		boolean ok = false;
		if (fileNo > 0 && fileNo <= structorizerFileMap.size()) {
			java.io.Closeable writer = structorizerFileMap.get(fileNo - 1);
			if (writer instanceof java.io.BufferedWriter) {
				((java.io.BufferedWriter)writer).write(data.toString());
				ok = true;
			}
		}
		if (!ok) { throw new java.io.IOException("fileWrite: Invalid file number or file not open for writing."); }
	}

	/**
	 * Writes the given value as character sequence to the file given by handle fileNo
	 * and appends a newline character or sequence as value separator.  
	 * Throws an error if the given handle is not associated to an open text output file.
	 * @param fileNo - file handle as obtained by fileCreate or fileAppend before
	 * @param data the object or value to be written to file.
	 */
	public static void fileWriteLine(int fileNo, java.lang.Object data) throws java.io.IOException
	{
		boolean ok = false;
		if (fileNo > 0 && fileNo <= structorizerFileMap.size()) {
			java.io.Closeable file = structorizerFileMap.get(fileNo - 1);
			if (file instanceof java.io.BufferedWriter) {
				((java.io.BufferedWriter)file).write(data.toString());
				((java.io.BufferedWriter)file).newLine();
				ok = true;
			}
		}
		if (!ok) { throw new java.io.IOException("fileWriteLine: Invalid file number or file not open for writing."); }
	}

}
